### PR TITLE
Zernike -> FWHM metric [DM-33115]

### PR DESCRIPTION
**Pull request by jfcrenshaw**
Added function to convert Zernike amplitudes (in microns) to their quadrature contributions to the PSF FWHM (in arcseconds). This is to address ticket [DM-33115](https://jira.lsstcorp.org/browse/DM-33115).

I wasn't exactly sure where to put this, so I put it in its own module/function in `utils`. I figured it would be good to open a pull request with the implementation in place, and someone more familiar with `ts_phosim` could tell me if and where to move it.

### CODE REVIEW

**Comment by teweitsai** [comment]
Please help:
1. Update the version history file for this new feature.
2. Put the license details in the new file if you want to put your new function in that file.
3. Put the parameters into the policy directory if possible.
4. Add the unit test to the new functions.
5. Make sure the pull request can pass the Jenkins tests.
6. Make sure your doc string follows the format of `numpy`.
7. I think there are some existed `enum` already when you check the available instrument. Try to use the existed `enum` if possible: https://github.com/lsst-ts/ts_wep/blob/develop/python/lsst/ts/wep/Utility.py#L42-L47

Thank you!

**Comment by jfcrenshaw** [comment]
@teweitsai I believe I have satisfied all of the points above, with one exception:

> 3. Put the parameters into the policy directory if possible.

I looked in the policy directory, but I'm not really sure what to do here. Are these all yaml files specifying defaults? Could you give me a little advice on what to do here?

Thanks for your review!

**Comment by teweitsai** [comment]
You can create a new `yaml` file or put the parameters into existed `yaml` file (if there is one that you think it is a good place). You can add comments to explain the new parameters. Thanks!

**Comment by jfcrenshaw** [comment]
Is this what you were thinking? There is only one parameter that can be defaulted.

https://github.com/jfcrenshaw/ts_phosim/blob/73f4f066e353cbc70dea2625636ae3b76a071dda/policy/convertZernikesToPsfWidthSetting.yaml#L1-L2

Or were you saying I should put the conversion factors in the yaml?
https://github.com/jfcrenshaw/ts_phosim/blob/73f4f066e353cbc70dea2625636ae3b76a071dda/python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py#L93-L170

**Comment by teweitsai** [comment]
I will put all the conversion factors in the `yaml` instead of hard-coding them. The `camType` is not the parameter but the argument of function (or the `optional` argument if you like). In addition, for the `enum` value, it is better to use the `name` instead of `value` in the `yaml` file. Otherwise, it is hard to understand what it is and what might be the available options.

**Comment by jfcrenshaw** [comment]
Okay the [policy file](https://github.com/jfcrenshaw/ts_phosim/blob/f7b109710d087a49f133c5a5dde5d9e812b29f8a/policy/convertZernikesToPsfWidthSetting.yaml) now contains the conversion factors.

Assuming the CI passes again, is this ready to merge?

**Comment by teweitsai** [comment]
No. You need to put the "Reviewers" to ask for the review of pull request. Then, after the approve, you can merge the pull request. What I did is called the "comment" instead of "review". 

In addition, it is better to rewrite the git history (to organize the comments) before asking the review or pull request.

**Comment by jfcrenshaw** [comment]
I cannot request reviews in the "Reviewers" field above because I do not have write access to this repo. But I have sent a slack message to @jbkalmbach asking him to review.

File: policy/convertZernikesToPsfWidthSetting.yaml
Comment (jbkalmbach) [inline policy/convertZernikesToPsfWidthSetting.yaml:3]: Does Josh have a notebook or something showing where these come from? Would be useful to have some documentation in case something changed and we need to do it again.

File: python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py
Comment (jbkalmbach) [inline python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py]: Functions in camel case should start with lower case letter.

File: python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py
Comment (jbkalmbach) [inline python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py:127]: Add `Raises` section explaining possible errors.

File: policy/convertZernikesToPsfWidthSetting.yaml
Comment (jbkalmbach) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: Would also be nice to make sure we have all cameras covered. How hard would it be to add ComCam, LsstFamCam coverage?

**Review by jbkalmbach — APPROVED** [review]
In general looks good, just a couple issues to fix and some questions.

**Comment by jbkalmbach** [comment]
One more comment now that I think about it some more. Do we want to change the file name to a more general name for Zernike analysis in case we want to add some more code in there to calculate metrics like we talked about at the last AOS meeting?

File: policy/convertZernikesToPsfWidthSetting.yaml
Comment (jfcrenshaw) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: @jmeyers314 can you answer these two questions? These are the values to convert the Zernike amplitudes from microns to their FWHM contributions.

File: python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py
Comment (jfcrenshaw) [reply python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py]: Fixed

File: python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py
Comment (jfcrenshaw) [reply python/lsst/ts/phosim/utils/ConvertZernikesToPsfWidth.py:127]: Done.

**Comment by jfcrenshaw** [comment]
> One more comment now that I think about it some more. Do we want to change the file name to a more general name for Zernike analysis in case we want to add some more code in there to calculate metrics like we talked about at the last AOS meeting?

1. Any suggestions for a name change? 
2. I think some metrics are already included in `OpdMetrology.py`, e.g. `calcPSSN`, `calcFWHMeff`, `calcDm5`. Maybe instead, this conversion should be moved there too? I'm just not sure how the `OpdMetrology` class gets used, since the docstring is pretty bare.

**Comment by jbkalmbach** [comment]
Yeah, it does look like if we make some sort of metrics utility we'll want those there too. So maybe let's just leave things the way you have them now and think about this some more after we start to converge on what the most useful metrics are as we go forward.

**Comment by jfcrenshaw** [comment]
@jbkalmbach sounds good. I will squash the history again, and then merge this addition.

I mentioned unifying the metrics as a future goal to Andy. He said he expects `ts_phosim` will be retired as we move away from `phosim`, and that the important pieces will need to be moved over to other packages. So maybe that is the stage where we should unify the metrics into some common framework.

File: policy/convertZernikesToPsfWidthSetting.yaml
Comment (jmeyers314) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: Here's a gist with the computation: https://gist.github.com/jmeyers314/38db2fd38b064be1e12c2992ea9b3e2d

Comment (jmeyers314) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: Looks like the values for AuxTel are a little bit different.  I wonder if this is confusion about the inner/outer pupil radii again.  I'm _pretty_ sure the values in the gist are correct, but somebody should double check me on those.

Given that the code is pretty short, and doesn't contain any new dependencies, maybe it'd be better to compute these coefficients on-the-fly rather than making a yaml file?  OTOH, the yaml file is nice as a quick reference for the relative influences of different Zernikes, so I'm not sure what's best.

Comment (teweitsai) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: I vote to compute the coefficients on-the-fly and the `yaml` file only needs to keep the radius of different telescopes.

Comment (jfcrenshaw) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: @teweitsai agreed, although I think we should just pull the inner and outer radii from [the yaml files in ts_wep](https://github.com/lsst-ts/ts_wep/tree/develop/policy/cwfs/instData). I'll start implementing this.

Comment (jfcrenshaw) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: I am using the [`Instrument` class from `ts_wep`](https://github.com/lsst-ts/ts_wep/blob/92110cfdaaacfcef9a9821cf4cb920be2c7d9487/python/lsst/ts/wep/cwfs/Instrument.py#L45) to pull out the radius and obscuration. You call `instrument.config(camType, dimOfDonutImgOnSensor)` to set up the instrument (`dimOfDonutImgOnSensor` is irrelevant for my purposes). 

This works fine for LsstCam, LsstFamCam, and ComCam. However, this fails for AuxTel because one of the optional arguments is `announcedDefocalDisInMm`, which must match one of the keys in the `offset` section of the instrument yaml file (for example, see the list of possible LsstCam offsets [here](https://github.com/lsst-ts/ts_wep/blob/develop/policy/cwfs/instData/lsst/instParam.yaml)).

This works for LsstCam, LsstFamCam, and ComCam because the default value is 1.5, and each of these instruments have 1.5mm as a potential offset in their yaml files. However, this fails for AuxTel, because the only option for AuxTel is 0.8mm (you can see this [here](https://github.com/lsst-ts/ts_wep/blob/develop/policy/cwfs/instData/auxTel/instParam.yaml)).

There are several ways to get around this:
- I could hard code that when `camType==AuxTel`, I should use `announcedDefocalDisInMm=0.8` when configuring the instrument.
- But If I am going to start hard coding parameters like this, why not just hardcode the radii and obscurations?
- Or I could just not use the `Instrument` class, and directly pull the values I want out of the yaml files.

Any thoughts on what I should do here? 

It seems funny to me that to even set up the instrument for AuxTel requires you to go read the yaml file and pass in the only offset value provided there. But maybe there is a good reason for that.

Comment (teweitsai) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: @jbkalmbach is trying to solve this here:
https://github.com/lsst-ts/ts_wep/pull/95

Hard-code is a bad idea.

Comment (jfcrenshaw) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: Great! I will wait on this until the new `Instrument` class/configs have settled.

Comment (jbkalmbach) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: I think `ts_wep` version 2.7 and higher should be able to handle what you want to do now.

Comment (jfcrenshaw) [reply policy/convertZernikesToPsfWidthSetting.yaml:3]: Great, thanks!

**Comment by jfcrenshaw** [comment]
Superseded by PR #72

**Pull request closed without merging**
